src/block.ts

```diff
@@ -0,0 +1,269 @@
+import Common from 'ethereumjs-common'
+import * as ethUtil from 'ethereumjs-util'
+import { BN, rlp } from 'ethereumjs-util'
+import { Transaction } from 'ethereumjs-tx'
+
+import { BlockHeader } from './header'
+import { Blockchain, BlockData, NetworkOptions } from './types'
+
+const Trie = require('merkle-patricia-tree')
+
+/**
+ * An object that represents the block
+ */
+export class Block {
+  public readonly header: BlockHeader
+  public readonly transactions: Transaction[] = []
+  public readonly uncleHeaders: BlockHeader[] = []
+  public readonly txTrie = new Trie()
+
+  private readonly _common: Common
+
+  /**
+   * Creates a new block object
+   *
+   *
+   * @param data - The block's data.
+   * @param opts - The network options for this block, and its header, uncle headers and txs.
+   */
+  constructor(
+    data: Buffer | [Buffer[], Buffer[], Buffer[]] | BlockData = {},
+    opts: NetworkOptions = {},
+  ) {
+    if (opts.common) {
+      if (opts.chain !== undefined || opts.hardfork !== undefined) {
+        throw new Error(
+          'Instantiation with both opts.common and opts.chain / opts.hardfork parameter not allowed!',
+        )
+      }
+
+      this._common = opts.common
+    } else {
+      const chain = opts.chain ? opts.chain : 'mainnet'
+      const hardfork = opts.hardfork ? opts.hardfork : null
+      this._common = new Common(chain, hardfork)
+    }
+
+    let rawTransactions
+    let rawUncleHeaders
+
+    if (Buffer.isBuffer(data)) {
+      // We do this to silence a TS error. We know that after this statement, data is
+      // a [Buffer[], Buffer[], Buffer[]]
+      const dataAsAny = rlp.decode(data) as any
+      data = dataAsAny as [Buffer[], Buffer[], Buffer[]]
+    }
+
+    if (Array.isArray(data)) {
+      // TODO: Pass the common object
+      this.header = new BlockHeader(data[0], opts)
```

Comment:
Is this difficult to directly pass the common object here instead of adding a placeholder TODO statement?

Reply:
(and instances below)

Reply:
It's not a problem here, but when initializing the transactions. The block defaults its hardfork to `null`, and `Transaction` doesn't support it.

I will pass the common object to the headers, and not to the transaction.

Reply:
I also two TODOs explaining the problems around Common, Transaction and block numbers